src/Microsoft.AspNetCore.Blazor.Server/Bots/RenderBatchWriter.cs

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Blazor.Rendering;
+using Microsoft.AspNetCore.Blazor.RenderTree;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Blazor.Server
+{
+    // TODO: We should consider *not* having this type of infrastructure in the .Server
+    // project, but instead in some new project called .Remote or similar, since it
+    // would also be used in Electron and possibly WebWorker cases.
+
+    /// <summary>
+    /// Provides a custom binary serializer for <see cref="RenderBatch"/> instances.
+    /// This is designed with both server-side and client-side perf in mind:
+    /// 
+    ///  * Array-like regions always have a fixed size per entry (even if some entry types
+    ///    don't require as much space as others) so the recipient can index directly.
```

Comment:
This doesn't seem true of https://github.com/aspnet/Blazor/pull/1090/files#diff-a60b475d488f645eaaa5af2613faaf8fR61

I'm following what you're doing in the code, but I'm not sure these comments are accurate.

Reply:
It might be worth listing which types and fixed size and which are variable size.

Reply:
`ArrayRange<RenderTreeDiff>` does output a fixed size per entry (each entry is a single int, which acts as a lookup into a different table of info).

I guess the comment in a bit ambiguous. The part I mean is "fixed size" is the array data itself, not any other data that it points to from there.